### DISCUSSION TITLE
Optimization with constraints

### DISCUSSION THREAD

**Discussion by abachma2** [snl-dakota/dakota#29] [Q&A] [answered]
I am using the `soga` method for optimization of a problem with a linear constraint, `x+y+z=100`. I see in the User's Manual that this method can handle this constraint form directly and in the Reference Manual that this method will "seek feasible solutions (i.e. solutions that satisfy the linear constraints), but may violate the constraints as they run. Linear constraints may also be violated, even when using an optimizer that itself strictly respects them". I have observed this in my runs (with Dakota considering cases that have `x+y+z=118.71`), but the final solution given does not fully meet the constraint (typically `98 <= x+y+z <= 102`). I even have a case in which `x+y+z=118.71` in the optimized solution found. Is there any way that I can more strictly enforce this linear constraint in my optimization problems?

Thanks in advance.

**Comment by jadamstephens** [comment]
I would try increasing the [constraint_penalty](https://dakota.sandia.gov/sites/default/files/docs/6.17.0-release/user-html/usingdakota/reference/method-soga-fitness_type-constraint_penalty.html). I'm forwarding your question to the `soga` developer to see if he has any other advice.

**Comment by jpeddy505** [comment, marked as answer]
In SOGA (and MOGA), equality constraints have a defined "thickness".  It is a violation amount that is considered allowed.  If I recall correctly, that parameter is exposed through the DAKOTA input spec though I don't know what the DAKOTA default value for that violation parameter is.  The internal JEGA value is 0 so can't be tightened up more than the default.  That's 1 avenue for tightening up your requirements on meeting the equality constraint.

Adam is correct that increasing the constraint_penalty will help if you are using a fitness assessor that has that parameter.

As far as the discussion of SOGA "handling" linear constraints, that refers to the fact that it can evaluate them internally given the coefficients and target instead of sending them out for evaluation through the DAKOTA interface.  It doesn't handle them in anything like the way that an LP would for example.

Another thing to consider is that choice of variational operators (mutation and crossover), in conjunction with your variable definitions may make it more difficult for the algorithm to respect equality constraints.  I can review your choices if you'd like and possibly offer suggestions.

If ultimately SOGA is unable to find solutions that meet the constraint to your satisfaction, you may consider seeding it with some x, y, z values that do meet the constraint.  This can be done using the flat_file initializer.  You run the risk of biasing the algorithm to the (likely superior) x,y,z values you provide but it does have some facilities to help maintain a healthy level of exploration even in the face of a few superior designs early in the process.

**Reply by jadamstephens** [reply]
Thanks, John. I don't think Dakota exposes that control for JEGA.

**Reply by jpeddy505** [reply]
In that case, the tolerance is already as tight as it can be and we'll have to find other ways to more accurately satisfy the constraint.

**Comment by abachma2** [comment]
Thank you both. I am currently working on tuning some of the parameters, including the constraint penalty, to see which set of parameters will do the best job in minimizing my problem and best meet the constraint. So it is good to know that I am on the right track by including the constraint penalty in my tuning. 

I am also trying to apply this constraint to a problem using MOGA and noticed that with a linear equality constraint I would only get one solution, instead of Pareto front. When I change it to a linear inequality constraint (`x+y+z>=100`) then I get multiple solutions. Should that be happening? I don't see the constraint penalty as a parameter for the MOGA method, so is there something else that I can vary to affect how well the equality constraint is met?

**Reply by jpeddy505** [reply]
It does not surprise me that you only get one solution.  With an equality constraint of this sort, it can be difficult to find multiple solutions that have exact equality or the "same level of equality" thus making 1 solution typically better than all the others.

Have you considered having only 2 variables and solving for the 3rd in your evaluation code to force the constraint?  For example, have DAKOTA choose x and y and then set z to 100 - (x+y)?

The MOGA does fitness assessment in a very different way and manages constraint violations in a very different way also.  There is no constraint_penalty for the fitness assessors used with it.

**Reply by abachma2** [reply]
That's a really good idea! I'll give that a try and see what happens.
